onchain/cairo/afk/src/bip340.cairo

```diff
@@ -126,14 +154,78 @@ pub fn verify(px: u256, rx: u256, s: u256, m: ByteArray) -> bool {
     // fail if is_infinite(R) || not has_even_y(R) || x(R) ≠ rx.
     !(Rx == 0 && Ry == 0) && Ry % 2 == 0 && Rx == rx
 }
+/// Represents a Schnorr signature
+struct SchnorrSignature {
+    s: felt252,
+    R: EcPoint,
+}
+
+/// Generates a key pair (private key, public key) for Schnorr signatures
+fn generate_keypair() -> (felt252, EcPoint) {
+    let generator: EcPoint = EcPointTrait::new(GEN_X, GEN_Y).unwrap();
+    let private_key: felt252 = 859825214214312162317391210310; // VRF needed
+    let public_key: EcPoint = generator.mul(private_key);
+    
+    (private_key, public_key)
+}
+
+/// Generates a nonce and corresponding R point for signature
+fn generate_nonce_point() -> (felt252, EcPoint) {
+    let generator: EcPoint = EcPointTrait::new(GEN_X, GEN_Y).unwrap();
+    let nonce: felt252 = 46952909012476409278523962123414653; // VRF needed
+    let R: EcPoint = generator.mul(nonce);
+    
+    (nonce, R)
+}
+
+/// Computes the challenge hash e using Poseidon
+fn compute_challenge(R: EcPoint, public_key: EcPoint, message: felt252) -> felt252 {
+    let (R_x, R_y): (felt252, felt252) = ec_point_unwrap(R.try_into().unwrap());
+    let (P_x, P_y): (felt252, felt252) = ec_point_unwrap(public_key.try_into().unwrap());
+    
+    PoseidonTrait::new()
+        .update(R_x)
+        .update(R_y)
+        .update(P_x)
+        .update(P_y)
+        .update(message)
+        .finalize()
+}
+
+/// Signs a message using Schnorr signature scheme
+fn sign(private_key: felt252, message: felt252) -> SchnorrSignature {
+    let (nonce, R) = generate_nonce_point();
+    let generator: EcPoint = EcPointTrait::new(GEN_X, GEN_Y).unwrap();
+    let public_key = generator.mul(private_key);
+    
+    let e = compute_challenge(R, public_key, message);
+    let s = nonce + mul_mod_p(private_key, e, ORDER);
+    
+    SchnorrSignature { s, R }
+}
+
+/// Verifies a Schnorr signature
+fn verify_sig(public_key: EcPoint, message: felt252, signature: SchnorrSignature) -> bool {
```

Comment:
We need also a conversion between this two types of Signature, like present in the others bip340 implement in our repo, and used in Nostr event:

  sig: Signature {
            r: 0xf1dac3f8d0d19767805ca85933bdf0e744594aeee04058eedaa29e26de087be9_u256,
            s: 0x144c4636083c7d0e3b8186c8c0bc6fa38bd9c6a629ec6e2ce5e437797a6e911c_u256
        }